code/android-core-library/src/main/java/com/adobe/marketing/mobile/rulesengine/download/RulesDownloader.java

```diff
@@ -0,0 +1,271 @@
+/*
+  Copyright 2022 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+ */
+
+package com.adobe.marketing.mobile.rulesengine.download;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
+
+import com.adobe.marketing.mobile.services.DeviceInforming;
+import com.adobe.marketing.mobile.services.HttpConnecting;
+import com.adobe.marketing.mobile.services.HttpMethod;
+import com.adobe.marketing.mobile.services.Log;
+import com.adobe.marketing.mobile.services.NetworkCallback;
+import com.adobe.marketing.mobile.services.NetworkRequest;
+import com.adobe.marketing.mobile.services.Networking;
+import com.adobe.marketing.mobile.services.caching.CacheEntry;
+import com.adobe.marketing.mobile.services.caching.CacheExpiry;
+import com.adobe.marketing.mobile.services.caching.CacheResult;
+import com.adobe.marketing.mobile.services.caching.CacheService;
+import com.adobe.marketing.mobile.util.StringUtils;
+import com.adobe.marketing.mobile.util.TimeUtils;
+import com.adobe.marketing.mobile.util.UrlUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.TimeZone;
+
+/**
+ * Facilitates the download and caching of rules from a url as well as an asset bundled with the app.
+ */
+public class RulesDownloader {
+    private static final String TAG = "RulesDownloader";
+
+    private static final int DEFAULT_CONNECTION_TIMEOUT_MS = 10000;
+    private static final int DEFAULT_READ_TIMEOUT_MS = 10000;
+
+    static final String HTTP_HEADER_IF_MODIFIED_SINCE = "If-Modified-Since";
+    static final String HTTP_HEADER_IF_NONE_MATCH = "If-None-Match";
+    static final String HTTP_HEADER_LAST_MODIFIED = "Last-Modified";
+    static final String HTTP_HEADER_ETAG = "ETag";
+
+    /**
+     * The cache name used for storing the downloaded results.
+     */
+    private final String cacheName;
+
+    private final Networking networkService;
+    private final CacheService cacheService;
+    private final DeviceInforming deviceInfoService;
+    private final RulesZipProcessingHelper rulesZipProcessingHelper;
+
+
+    public RulesDownloader(@NonNull final String cacheName,
```

Comment:
Can you make this non public and add another constructor with just the cacheName? We can retrieve the other instances from `ServiceProvider`.